Modify rule S2694: Add code example for local classes (#3748)
SonarSource/rspec: rules/S2694/java/rule.adoc

* an inner class can only be instantiated within the context of an instance of the outer class.
* a nested (``++static++``) class can be instantiated independently of the outer class.

== How to fix it

> There are two scenarios where the issue can be raised:

Reviewer: How about
```suggestion
There are two scenarios in which this rule will raise an issue:
```

1. On a `inner class`: make it ``++static++``.

> 2. On a `local class`: extract it as a `static inner class`.

Reviewer: I think marking things like "inner class" as `code` is confusing here. How about putting non-code phrases that you want to highlight into _italics_ instead?

=== Code examples

==== Noncompliant code example

> Inner class that doesn't use the outer class reference should be static.

Reviewer: ```suggestion
Inner classes that doesn't use the outer class reference should be static.
```

}
----

Local class that doesn't use the outer class reference should be extracted as a static inner class.

==== Noncompliant code example

[source,java,diff-id=2,diff-type=noncompliant]
----
public class Foo {
  public Foo() {

>     class Bar { // Compliant

Reviewer: Should be a `Noncompliant` instead of `Compliant` comment here

      void doSomething() {
        // ...
      }
    }
    new Bar().doSomething();
  }

  public void method() {

>     class Baz { // Compliant

Reviewer: Should be a `Noncompliant` instead of `Compliant` comment here

      void doSomething() {
        // ...
      }
    }
    new Baz().doSomething();
  }
}
----

==== Compliant solution

[source,java,diff-id=2,diff-type=compliant]
----
public class Foo {
  public Foo() {
    new Bar().doSomething();
  }

  public void method()  {
    new Baz().doSomething();
  }

  static class Bar { // Compliant
    void doSomething() {
      // ...
    }
  }

>   static class Baz { // Compliant

Reviewer: Perhaps make the classes private to avoid opening up access too much?
